src/mlpack/core/optimizers/sgdr/sgdr.hpp

```diff
@@ -0,0 +1,256 @@
+/**
+ * @file sgdr.hpp
+ * @author Marcus Edel
+ *
+ * Definition of the Stochastic Gradient Descent with Restarts (SGDR) as
+ * described in: "SGDR: Stochastic Gradient Descent with Warm Restarts" by
+ * I. Loshchilov et al and the Snapshot ensembles technique described in:
+ * "Snapshot ensembles: Train 1, get m for free" by G. Huang et al.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_SGDR_SGDR_HPP
+#define MLPACK_CORE_OPTIMIZERS_SGDR_SGDR_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include <mlpack/core/optimizers/minibatch_sgd/minibatch_sgd.hpp>
+#include <mlpack/core/optimizers/sgd/update_policies/momentum_update.hpp>
+#include <mlpack/core/optimizers/sgdr/cyclical_decay.hpp>
+#include <mlpack/core/optimizers/sgdr/snapshot_ensembles.hpp>
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * This class is based on Mini-batch Stochastic Gradient Descent class and
+ * simulates a new warm-started run/restart once a number of epochs are
+ * performed this class also implements the Snapshot ensembles technique.
+ *
+ * For more information, please refer to:
+ *
+ * @code
+ * @article{Loshchilov2016,
+ *   title   = {{SGDR:} Stochastic Gradient Descent with Restarts},
+ *   author  = {Ilya Loshchilov and Frank Hutter},
+ *   journal = {CoRR},
+ *   year    = {2016}
+ * }
+ * @endcode
+ *
+ * @code
+ * @inproceedings{Huang2017,
+ *   title     = {Snapshot ensembles: Train 1, get m for free},
+ *   author    = {Gao Huang, Yixuan Li, Geoff Pleiss, Zhuang Liu,
+ *                John E. Hopcroft, and Kilian Q. Weinberger},
+ *   booktitle = {Proceedings of the International Conference on Learning
+ *                Representations (ICLR)},
+ *   year      = {2017}
+ * }
+ * @endcode
+ *
+ * @tparam DecomposableFunctionType Decomposable objective function type to be
+ *         minimized.
+ * @tparam UpdatePolicyType Update policy used during the iterative update
+ *         process. By default the vanilla update policy
+ *         (see mlpack::optimization::VanillaUpdate) is used.
+ * @tparam DecayPolicyType Decay policy used during the iterative update
+ *         process to adjust the step size (CyclicalDecay or SnapshotEnsembles).
+ */
+template<
+    typename DecomposableFunctionType,
+    typename UpdatePolicyType = MomentumUpdate,
+    typename DecayPolicyType = CyclicalDecay
```

Comment:
It looks like `DecayPolicyType` can only be `SnapshotEnsembles` or `CyclicalDecay`, and the implementation is different in either case.  So do you think it would be better to simply have two different classes?  I guess that might obviate the need for `DecayPolicyType` in general too; not sure.